### ISSUE TITLE
subprojects ignore `shared_module`'s `build_by_default : false` option when `install : true` and build regardless

### ISSUE THREAD

**Issue by lf-** [mesonbuild/meson#13498]
**Describe the bug**
If you set `build_by_default : false` in a subproject, one would expect the target to not be built by default. I later discovered this is because `install : true`, which is not documented to be related to whether it will be be built by default.

The use case that is leading to wanting this is that this is a clang-tidy plugin that I would like to build only on demand when you actually run the `clang-tidy` custom target, and it is in a subproject because I would like to be able to also pass it in externally to easily use Nix build caching. Actually it would be even more ideal if the subproject were not even *configured* if the dependent target is not demanded, since there is no reason to need it to be, but I would also accept skipping the build.

**To Reproduce**

https://git.jade.fyi/jade/meson-repro-20240731

or, reproduce the below file structure:

```
meson/repro » tree .
.
├── meson.build
└── subprojects
    └── subproj
        ├── lib.c
        └── meson.build

3 directories, 3 files

meson/repro » cat meson.build
project('repro')
subproject('subproj')
meson/repro » cat subprojects/subproj/meson.build
project('lix-clang-tidy', ['c'])

sources = files('lib.c')
module = shared_module('lix-clang-tidy', sources, install : true, build_by_default : false)

meson/repro » cat subprojects/subproj/lib.c
```

Repro:

```
meson/repro » meson setup build
....

meson/repro » rg 'build all' build/build.ninja
115:build all: phony meson-test-prereq meson-benchmark-prereq subprojects/subproj/liblix-clang-tidy.so
```

Observe that `liblix-clang-tidy.so` is incorrectly included in `all` in spite of being explicitly set to not be built by default.

**Expected behavior**

I expect the subproject target to not be built by default if it is not set to be built by default.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? Nope
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.): Arch Linux
* what Python version are you using e.g. 3.8.0: 3.12.4
* what `meson --version`: 1.5.1
* what `ninja --version` if it's a Ninja build: don't even have to run ninja to find this problem, but 1.12.1

**Comment by eyalitki** [comment]
This is a larger bug, as it isn't directly tied to subprojects. Fix will be posted soon.

**Comment by eyalitki** [comment]
Well, it seems I was overly optimistic, and I couldn't find a step in which the "install" phase can trigger the build of targets that were skipped during the "compile" phase. While this might be possible, it is beyond my understanding of the project.

This also means that "build_by_default" currently only makes sense when install is "false". If this is a concise decision, the documentation (and comments in the code) should probably be updated accordingly.

**Comment by eli-schwartz** [comment]
That is broadly correct, yes. In the distant past, we didn't build those files and instead raised incomprehensible tracebacks during `meson install` because required files weren't found.

There is a related issue somewhere but I can't remember where off the top of my head.

Note: it's fairly common for build systems to provide a default target (usually called "all" that compiles all build rules which are needed for, well, *installing* the results. The current meson behavior is, in a sense, not surprising.

(In another sense, it's arguably surprising that you can try setting `build_by_default: false` without triggering a warning/error "this setting has no effect because the target is installed.)

**Comment by eyalitki** [comment]
Ack, understood.
1. Should I instead provide a pull request to update the documentation accordingly?
2. Should said pull request also add a warning about this setting having no effect?

**Comment by eyalitki** [comment]
I tracked the following related issues: #4370, #12530 and #13453.

**Comment by bonzini** [comment]
@eyalitki yes doing both of those in the pull request would be a good idea

**Comment by bonzini** [comment]
@eyalitki I cannot find the PR [here](https://github.com/mesonbuild/meson/pulls/eyalitki), can you link it?

**Comment by bonzini** [comment]
For the specific case of the clang-tidy plugin, 1.12 lets you invoke the subproject with `native: true` and then it won't be installed at all.
